### ISSUE TITLE
True Meaning of Argument "imgsz"?

### ISSUE THREAD

**Issue by A1exan10er** [ultralytics/ultralytics#2546]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello ultralytics team,
I have a question regarding setting the value of "imgsz" for training.
By default, its value is 640 and some people will change it to 1280 when detecting small objects, like potholes on road.
I checked the explanations at "arguments for training" on https://docs.ultralytics.com/modes/train/?h=arguments#arguments and it says 

imgsz | 640 | size of input images as integer or w,h
-- | -- | --

According to this explanation, "imgsz" is the size of input images. Then if images from training dataset is 2160x3840 (4k resolution), should this value become 3840, since the input is this value?
Or can it still pick 640?
What would be the difference between choosing 640, 1280 or even higher values?
To be precise, what does this argument really do?
If I train with 640 and try predicting some potholes on a 2160x3840 image, would it have better performance than using 1280 or higher value?

I'm confused here. Hope someone can explain this part. Thanks in advance!

Best regards,
Tianyu

**Comment by ExtReMLapin** [comment]
If you're image is 2160x2160 it will downscale it (without augmentations) to 640x640

It takes the longest edge, and fit scales it to 640 and fits it in a box.

Long story short, you may have huge resolution dataset but if you don't have tiny details on it, you can keep it the way it is.

You might also for tiny details consider using the P2 model which enable tiny details (<8px iirc)

**Comment by A1exan10er** [comment]
> If you're image is 2160x2160 it will downscale it (without augmentations) to 640x640
> 
> It takes the longest edge, and fit scales it to 640 and fits it in a box.
> 
> Long story short, you may have huge resolution dataset but if you don't have tiny details on it, you can keep it the way it is.
> 
> You might also for tiny details consider using the P2 model which enable tiny details (<8px iirc)

Thank you for your reply!

https://stackoverflow.com/questions/75268393/yolov8-how-does-it-handle-different-image-sizes
I also found one explanation here, and it meant "imgsz" reshape input image.

What I'm not sure is, during training, if my dataset image has a 16:9 aspect ratio, for example 1920x1080 (pixels). If "imgsz" was set to 640, would it turn the input image into a square form? If so, would it be bad for training because some small objects (potholes) might be shrunk or deformed?

And yes in my dataset there're lots of small details, but not too small, guess most of them are several hundred (>200) pixels. And also some larger objects, not too large, like laptops and notebooks etc. I think in my case, YOLO can deal with it well.

You also mentioned "P2, P5, P6, P7 layers", do you mean they're different layers used in YOLOv8? I've not yet checked its structure, so not quite understand this. I think you mean there're different layers in YOLOv8, and some work well at detecting large objects, and some work well at detecting small ones, and P7 layer is capable of detecting both?

**Comment by ExtReMLapin** [comment]
> If so, would it be bad for training because some small objects (potholes) might be shrunk or deformed?

Not, it will add gray around it, it keep ratio as it is.

You can take a look at the generated `train_batch0` files

![](https://i.imgur.com/I455zO8.png)
it adds gray around it.



>You also mentioned "P2, P5, P6, P7 layers", do you mean they're different layers used in YOLOv8? I've not yet checked its structure, so not quite understand this. I think you mean there're different layers in YOLOv8, and some work well at detecting large objects, and some work well at detecting small ones, and P7 layer is capable of detecting both?

Long story short, P2 is to detect very smalls objects, P6, P7 biggers objects.

There is better documentation on that on YoloV5 , but you can easily port yaml architectures to YoloV8 (here)

See glen answer here : https://github.com/ultralytics/yolov5/discussions/9862

For example, the p6 layer only meant to exist for when you train a dataset with size = 1280 instead of 640, see this table :  

https://github.com/ultralytics/yolov5#pretrained-checkpoints

YoloV5n6 is just YoloV5n with a P6 layer, (check the diff yourself in models/ and models/hub in yolov5 repo) ans the only thing he also changes is the image resolution.

I'm clearly not an ML expert, I'm just guessing

**Comment by glenn-jocher** [comment]
@A1exan10er hello Tianyu!

Regarding your question about the "imgsz" argument for training, setting it to 640 will not disturb the aspect ratio of images that have a 16:9 aspect ratio, like 1920x1080 pixels. Instead, the input image will be padded with gray to make it a square while keeping its aspect ratio. This will not be bad for training because the padded space is not part of the image, so it will not affect the detection of small objects (potholes).

Additionally, P2, P5, P6, and P7 are different layers used in YOLOv8. These layers have different detection capabilities for different object sizes. P2 is designed to detect very small objects, while P6 and P7 are for bigger objects. If you have small details in your dataset, you may consider using the P2 model. On the other hand, if there are some larger objects in your dataset, P6 and P7 models can work well.

Feel free to let us know if you have more questions or concerns.

Best regards, 
Glenn Jocher

**Comment by dandv** [comment]
I'm also unclear on this option.

1. Why bother wondering whether the aspect ratio would be preserved, or grey added, when `imgsz` can be specified as `w,h` instead of one number?
2. Still, when specifying `imgsz='1280,736'`, ultralytics apparently flips the dimensions because I see this in the verbose log:

        video 1/1 (frame 2/106529) myvideo.mp4: 416x736 (no detections), 56.0ms
        
   Must the dimensions be specified as `height,width` instead?
3. Even when I pass `imgsz='736,1280'`, both already multiples of "max [stride](https://deepai.org/machine-learning-glossary-and-terms/stride)" 32, I still get the warning that imgsz was updated, when it was not:

        WARNING ⚠ imgsz=(736, 1280) must be multiple of max stride 32, updating to [736, 1280]
        video 1/1 (frame 1/20547) myvideo.mp4: 736x1280 (no detections), 165.4ms

   Is this a bug?

**Comment by dandv** [comment]
**UPDATE** I picked up the "w,h" syntax from the OP, but it is no longer listed at the https://docs.ultralytics.com/modes/train/?h=arguments#arguments link mentioned by @A1exan10er.

The [predict arguments](https://docs.ultralytics.com/modes/predict/#inference-arguments) list "h, w" indeed.

**Comment by Y-T-G** [comment]
@dandv Most of the time in image processing related things, image shape is in `h,w` format. That's the case even with OpenCV. Ultralytics also uses the same format.

**Comment by Y-T-G** [comment]
> 3. Even when I pass `imgsz='736,1280'`, both already multiples of "max [stride](https://deepai.org/machine-learning-glossary-and-terms/stride)" 32, I still get the warning that imgsz was updated, when it was not:
>    ```
>     WARNING ⚠ imgsz=(736, 1280) must be multiple of max stride 32, updating to [736, 1280]
>     video 1/1 (frame 1/20547) myvideo.mp4: 736x1280 (no detections), 165.4ms
>    ```

It's supposed to be either an int or a tuple, not string, as mentioned in the [docs](https://docs.ultralytics.com/modes/predict/#inference-arguments).

**Comment by binhere** [comment]
> Regarding your question about the "imgsz" argument for training, setting it to 640 will not disturb the aspect ratio of images that have a 16:9 aspect ratio, like 1920x1080 pixels. Instead, the input image will be padded with gray to make it a square while keeping its aspect ratio. This will not be bad for training because the padded space is not part of the image, so it will not affect the detection of small objects (potholes).

@glenn-jocher , I wonder if model YOLOv8 predicts an image of shape (1080; 1920) with `imsz=640`, does it keep my image's aspect ratio to the new shape with a longer edge of 640px, or will my image become (640; 640)? Does my input image have distortion?

**Comment by glenn-jocher** [comment]
For YOLOv8 with `imgsz=640`, your 1080x1920 image will be resized to maintain aspect ratio (longest side scaled to 640px) and padded with gray to create a 640x640 square input. No distortion occurs - the original aspect ratio is preserved. You can verify this behavior by checking the generated `train_batch*.jpg` images during training, which show the actual preprocessed inputs. For more details, see the Image Preprocessing section in the YOLOv8 documentation.

**Comment by enescicek** [comment]
Hello @glenn-jocher 

If I'm working with large and detailed images, would setting the imgsz parameter as large as possible be beneficial?
The images I'm working with contain small details that I need to detect accurately. I'm doing segmentation work.

**Comment by glenn-jocher** [comment]
Larger `imgsz` can improve segmentation of fine details **only if** those details get lost when your images are resized down, but “as large as possible” usually isn’t optimal because VRAM/latency grow quickly and you may be forced into a much smaller `batch`, which can hurt training stability; a good workflow is to train/val a couple of sizes (e.g. 640 → 960/1024 → 1280) and pick the smallest that meets your mask-quality target, and if you still need more detail consider tiling/cropping your large images instead of pushing `imgsz` further. Input resizing keeps aspect ratio and uses letterbox padding (no distortion) as implemented in [`LetterBox`](https://docs.ultralytics.com/reference/data/augment/#ultralyticsdataaugmentletterbox), and your requested size may be rounded to a stride-multiple via [`check_imgsz`](https://docs.ultralytics.com/reference/utils/checks/). Since this issue is closed/stale, please open a new issue (or a GitHub Discussion) with your image size, smallest object size, GPU, and a couple of sample images/masks so we can give a more specific recommendation.
